### PR TITLE
Add setpoint derivative feed forward term to PID controller

### PR DESCRIPTION
**Pull request by DzikuVx**
As part of this PR:

* Dterm is always computed from measurement
* Control Derivative is computed from filtered setpoint
* Removed Dterm FIR filter
* Removed Dterm notch
* Omnibus F3 removed from release list

### CODE REVIEW

**Comment by digitalentity** [comment]
So, this is not a feed-forward itself, it's a multiplier for one of the components of the D-term

**Comment by DzikuVx** [comment]
No it's not. And it's a shame BF called that a feed forward. It's not done,
the goal is to comoute dterm only from measurement.
From the math level it's the same as now, only tuning is simpler since
instead of setpoint weight you separately tune "damping" and acceleration
from the stick movement

wt., 28 kwi 2020, 17:42 użytkownik Konstantin Sharlaimov <
notifications@github.com> napisał:

**Comment by digitalentity** [comment]
Maybe we should then separate D/setpoint and D/measurement into two PID parameters and deprecate setpoint weight?

We'll end up with P, I, D/setpoint, D/meas and FF, may be a lot, but it will in fact may tuning simpler. D/setpoint and FF will only affect reaction to the stick.

**Comment by DzikuVx** [comment]
That would mean we will need 5th gain in each bank, new frames for MSP, Configurator support and so on. 
So, in general I agree that would be good, but quite complicated. As long as possible I want to avoid adding a 5th gain to each PID bank, even if naming will be not quite right.
Also, pilots on MC are used to BF FF and that another reason not to overcomplicate.

**Comment by digitalentity** [comment]
Makes sense, but can we please not call it "feed-forward" to avoid confusion with real FF-term on fixed wings?

**Comment by DzikuVx** [comment]
DS maybe?

śr., 29 kwi 2020, 13:32 użytkownik Konstantin Sharlaimov <
notifications@github.com> napisał:

**Review by digitalentity — DISMISSED** [review]

**Comment by DzikuVx** [comment]
First flight tests completed and result are very appealing. There is less vibrations due to better behaving Dterm that now is measured only from measurement. The amount of control on fast strick movements was better than before. Unfortunatley I do not have a log with CDterm recorded.

image

DVR: https://youtu.be/tFXFGss8aFE

It still needs some check but seems like a good feature

File: src/main/flight/pid.h [hunk 1]
```diff
@@ -45,6 +45,7 @@ FP-PID has been rescaled to match LuxFloat (and MWRewrite) from Cleanflight 1.13
 #define FP_PID_RATE_P_MULTIPLIER    31.0f
 #define FP_PID_RATE_I_MULTIPLIER    4.0f
 #define FP_PID_RATE_D_MULTIPLIER    1905.0f
+#define FP_PID_RATE_D_FF_MULTIPLIER   7270.0f
```
Comment (digitalentity) [inline src/main/flight/pid.h:48]: Would it make sense to have the same multiplier as D-term? It's effectively the same thing, wouldn't it make sense to use the same scale?

File: src/main/flight/pid.c [hunk 2]
```diff
@@ -496,13 +469,15 @@ void FAST_CODE NOINLINE updatePIDCoefficients(float dT)
...
             pidState[axis].kFF = pidBank()->pid[axis].FF / FP_PID_RATE_FF_MULTIPLIER * tpaFactor;
+            pidState[axis].kCD = 0.0f;
             pidState[axis].kT  = 0.0f;
         }
         else {
             const float axisTPA = (axis == FD_YAW) ? 1.0f : tpaFactor;
             pidState[axis].kP  = pidBank()->pid[axis].P / FP_PID_RATE_P_MULTIPLIER * axisTPA;
             pidState[axis].kI  = pidBank()->pid[axis].I / FP_PID_RATE_I_MULTIPLIER;
             pidState[axis].kD  = pidBank()->pid[axis].D / FP_PID_RATE_D_MULTIPLIER * axisTPA;
+            pidState[axis].kCD = pidBank()->pid[axis].FF / FP_PID_RATE_D_FF_MULTIPLIER * axisTPA;
```
Comment (digitalentity) [inline src/main/flight/pid.c:480]: This is soooo uuuuugly :grimacing:

File: src/main/flight/pid.h (hunk 1 above)
Comment (DzikuVx) [reply src/main/flight/pid.h:48]: I want to keep this thing 1:1 on values with Betaflight. That will be simpler to migrate from BF to INAV

File: src/main/flight/pid.c (hunk 2 above)
Comment (DzikuVx) [reply src/main/flight/pid.c:480]: I will not make separate `CD` entry :) I dont want to have new MSP frames

File: src/main/flight/pid.h (hunk 1 above)
Comment (digitalentity) [reply src/main/flight/pid.h:48]: Makes sense then.

File: src/main/flight/pid.c (hunk 2 above)
Comment (digitalentity) [reply src/main/flight/pid.c:480]: Not yet at least :smile:

**Comment by digitalentity** [comment]
Happy to hear. Finally we're getting rid of `dterm_setpoint_weight` in favor of more intuitive tuning.

**Comment by Avegawanderer** [comment]
Maybe let's call the derivative of setpoint not DS but DFF (Differential Feed-Forward, or Differentiated Feed Forward) People have got used to "Feed Forward" name and it would be easier to explain that Betaflight's FF is actually DFF in iNav and the real feed-forward is just FF in iNav. At least both FF and DFF names describe their impact for the system. DS does not.

**Comment by digitalentity** [comment]
IMO "feed-forward" has nothing to do with how this works (neither in INAV nor in BF). 
The behavior is exactly about taking a derivative of setpoint (control) and propagating that straight to the motors. This is exactly what `dterm_setpoint_weight` does, just having a dedicated gain, decoupled from derivative of measurement (gyro).

**Comment by Avegawanderer** [comment]
Yes, what is called FF in Betaflight and iNav is different from true FF in a control system theory. But speaking about iNav, FF for fixed-wings simply propagates setpoint (control) to output (servos). While such FF is not used for multirotors, it's derivative is used, and, as I understand (please correct me if I'm wrong) Pawel in this PR brings it to PID controller settings, dropping dterm_setpoint_weight. The question is to name it properly. My proposal is to keep FF as it is for FW and name new component DFF instead of DS. It directly explains what it does (takes FF (=setpoint) and gets it's derivative). The goal is to use familiar FF name.

**Comment by digitalentity** [comment]
I see, so the idea is to avoid adding a new term "setpoint" to prevent user confusion. @DzikuVx WDYT?

**Comment by DzikuVx** [comment]
Actually I think that `Control Derivative` which is the name used in the code right now is exactly what this functionality is.
I'm even thinking about renaming `D` in configurator to `Gyro Derivative` so there is zero confusion what is what

**Pull request merged**